src/vs/base/common/cancellation.ts

```diff
@@ -24,15 +42,40 @@ const shortcutEvent = Object.freeze(function (callback, context?): IDisposable {
 
 export namespace CancellationToken {
 
+	/**
+	 * A token that will never be canceled.
+	 */
 	export const None: CancellationToken = Object.freeze({
 		isCancellationRequested: false,
-		onCancellationRequested: Event.None
+		onCancellationRequested: Event.None,
+		throwIfCancellationRequested() { },
 	});
 
+	/**
+	 * A token that is already canceled.
+	 */
 	export const Cancelled: CancellationToken = Object.freeze({
 		isCancellationRequested: true,
-		onCancellationRequested: shortcutEvent
+		onCancellationRequested: shortcutEvent,
+		throwIfCancellationRequested() {
+			throw canceledError();
+		},
 	});
+
+	/**
+	 * Returns an error that may be used to communicate cancellation.
+	 */
+	export function canceledError() {
+		return canceled();
```

Comment:
:-1: I don't see the value in re-exporting this

Reply:
It isn't publicly exported otherwise. The original method is only exported from an internal module.

Reply:
curious to know what your definition of an 'internal module' is?

Reply:
I can't reach it from outside the vscode-jsonrpc package. As an NPM consumer of the vscode-jsonrpc package, I need these two methods that I'm exporting, and they are not exported at the package level today.
Further, rather than export them publicly at the package level where they are currently at, these two members (IMO) ought to be exported directly from the types that they work with so they are more discoverable. 
Ultimately, I'd like to see CancellationToken pulled out into its own NPM package and consumed from vscode-jsonrpc. Until then, I'm adding to the API here as if it were a self-contained API.

Reply:
Ah, you're right: this is the wrong repo.

My goal: I'm writing an NPM package for [multiplexing streams](https://www.npmjs.com/package/nerdbank-streams) to work as a companion to your vscode-jsonrpc package. I already have it defined in a .NET library and of course I want cancellation support. I certainly do not want to redefine cancellation tokens as your package has already done it. My package does not require JSON-RPC itself, but it depends on vscode-jsonrpc the package in order to get cancellation tokens. But I need the functionality that this PR adds (albeit to the wrong repo) to be complete.

So it seems I should close this PR and open one against the right repo. And if you'd entertain the idea, perhaps I can start a new repo with all your cancellation token code in it as a new package, then send a PR to this and/or the vscode-languageserver-node repo to consume the new one so you have just one code base for cancellation tokens. 

Thoughts?

Reply:
Yeah, when you only need the cancellation logic you should create a separate npm-module with a copy of our code. That makes good sense. The json-rpc package should be able to adopt that with relative ease, for vscode itself that's a little more complex due to the way how we package our code.

Reply:
Hmm... In looking for a good NPM package name for isolated cancellation token functionality, I ran across an existing [cancellationtoken](https://www.npmjs.com/package/cancellationtoken) package. While its API isn't identical to yours nor .NET's, it looks very good. It's a little more developed (it already has the methods/properties I was adding here) and more, using terminology that is natural for node.js consumers rather than the .NET crowd, but closely enough that a .NET user could find his/her way around.

I'm going to try switching to using the existing cancellationtoken package for my purposes. Do you think the vscode-jsonrpc is locked into its current API for backward compat reasons?

Reply:
(maybe we could leverage TypeScript's union types to allow for either type of cancellation token.

Reply:
json-rpc is not locked into it. We can if we want move to a 4.x version. 

Or types will not avoid the breakage since callback signature have a token. And widening the type of a param is breaking.